### PR TITLE
feat(module:image): add nz-image component

### PR DESCRIPTION
**Pull request by stygian-desolator**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by stygian-desolator** [comment]
@hsuanxyz please mark as draft, thanks~

**Comment by stygian-desolator** [comment]
@hsuanxyz pls help me review this, thx~

File: components/image/demo/preloading.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+---
+order: 7
+title:
+  zh-CN: 预先加载
+  en-US: Preloading
+---
+
+## zh-CN
+
+添加 `nzPreload` 可以在服务端渲染下预先加载该图片。
```
Comment (hsuanxyz) [inline components/image/demo/preloading.md:10]:
```suggestion
添加 `nzPreload` 可以在服务端渲染下添加 [preload](https://developer.mozilla.org/en-US/docs/Web/HTML/Preloading_content)
```

File: components/image/demo/preloading.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+## en-US
+
+Adding `nzPreload` will preload the image under server-side rendering.
```
Comment (hsuanxyz) [inline components/image/demo/preloading.md:14]:
```suggestion
Using `nzPreload` will add (preload)[https://developer.mozilla.org/en-US/docs/Web/HTML/Preloading_content]  under server-side rendering.
```

File: components/image/doc/index.en-US.md [hunk 3]
```diff
@@ -31,6 +31,18 @@ import { NzImageModule } from 'ng-zorro-antd/image';
...
+
+| Property | Description | Type | Default | Global Config |
+| --- | --- | --- | --- | --- |
+|nzSrc | URL | `string` | - | |
+|nzAlt | Alt | `string` | - | |
+|nzWidth | Width | `number\|string` | `auto` | |
+|nzHeight | Height | `number\|string` | `auto` | |
+|nzLoader | Loader | `NzImageLoader` | `defaultLoader` | ✅ |
+|nzOptimize | Whether to optimize image loading | `boolean` | `false` | ✅ |
+|nzPreload | Whether to preload image | `boolean` | `false` | ✅ |
```
Comment (hsuanxyz) [inline components/image/doc/index.en-US.md:44]:
```suggestion
|nzPreload | Whether to add [preload](https://developer.mozilla.org/en-US/docs/Web/HTML/Preloading_content) (only SSR) | `boolean` | `false` | ✅ |
```

File: components/image/doc/index.zh-CN.md [hunk 4]
```diff
@@ -32,6 +32,19 @@ import { NzImageModule } from 'ng-zorro-antd/image';
...
+
+| 参数        | 说明                               | 类型             | 默认值 | 支持全局配置  |
+| --- | --- | --- | --- | --- |
+|nzSrc | url | `string` | - | |
+|nzAlt | alt | `string` | - | |
+|nzWidth | 宽度 | `number\|string` | `auto` | |
+|nzHeight | 高度 | `number\|string` | `auto` | |
+|nzLoader | 图片加载器 | `NzImageLoader` | `defaultLoader` | ✅ |
+|nzOptimize | 是否优化图片加载 | `boolean` | `false` | ✅ |
+|nzPreload | 是否预加载图片 | `boolean` | `false` | ✅ |
```
Comment (hsuanxyz) [inline components/image/doc/index.zh-CN.md:45]:
```suggestion
|nzPreload | 是否添加 [preload](https://developer.mozilla.org/en-US/docs/Web/HTML/Preloading_content) | `boolean` | `false` | ✅ |
```

File: components/image/image.component.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+  @Input() nzAlt: string = '';
+  @Input() nzWidth: string | number = 'auto';
+  @Input() nzHeight: string | number = 'auto';
+  @Input() @WithConfig() nzLoader: NzImageLoader = defaultLoader;
+  @Input() @InputBoolean() @WithConfig() nzOptimize: boolean = false;
+  @Input() @InputBoolean() @WithConfig() nzPreload: boolean = false;
+  @Input() @WithConfig() nzOptimizeSizes: number[] = [16, 32, 48, 64, 96, 128, 256, 384, 640, 750, 828, 1080, 1200, 1920, 2048, 3840];
+
+  get width(): number {
+    return typeof this.nzWidth === 'number' ? this.nzWidth : parseInt(this.nzWidth, 10);
```
Comment (hsuanxyz) [inline components/image/image.component.ts:55]: The `nzWidth` default value is `auto` will be return `NaN` (`parseInt('auto', 10)`).

File: components/image/image.component.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  private preload(): void {
+    if (isPlatformServer(this.platformId)) {
+      const linkNode = this.render.createElement('link') as HTMLLinkElement;
+      linkNode.rel = 'preload';
+      linkNode.as = 'image';
+      linkNode.imageSizes = this.sizes;
+      linkNode.imageSrcset = this.srcSet;
+      linkNode.href = this.src;
+      const headNode = this.document.getElementsByTagName('head')[0] as HTMLHeadElement;
+      this.render.appendChild(headNode, linkNode);
```
Comment (hsuanxyz) [inline components/image/image.component.ts:110]:
```suggestion
     this.renderer.appendChild(this.document.head, linkNode);

```

File: components/image/image.component.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    private cdr: ChangeDetectorRef,
+    public nzConfigService: NzConfigService,
+    @Inject(PLATFORM_ID) private platformId: NzSafeAny,
+    @Inject(DOCUMENT) private document: NzSafeAny
+  ) {
+    this.nzConfigService
+      .getConfigChangeEventForComponent(NZ_CONFIG_MODULE_NAME)
+      .pipe(takeUntil(this.destroy$))
+      .subscribe(() => {
+        this.cdr.markForCheck();
```
Comment (hsuanxyz) [inline components/image/image.component.ts:79]: anything forgot here?

File: components/image/doc/index.en-US.md [hunk 8]
```diff
@@ -69,3 +81,13 @@ Other attributes [<img\>](https://developer.mozilla.org/en-US/docs/Web/HTML/Elem
 | prev(): void | Previous image |
 | next(): void | Next image |
 | close(): void | Close image preview |
+
+### NzImageLoader
```
Comment (hsuanxyz) [inline components/image/doc/index.en-US.md:85]: More details are needed here. And we also need the built-in loaders.

File: components/image/image.component.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+      .join(', ');
+  }
+
+  private convertWidths(width: number, optimizeSizes: number[]): number[] {
+    const allSizes = [...optimizeSizes].sort((a, b) => a - b);
+    return [
+      ...new Set(
+        // 2x scale is sufficient
+        // https://blog.twitter.com/engineering/en_us/topics/infrastructure/2019/capping-image-fidelity-on-ultra-high-resolution-devices.html
+        [width, width * 2].map(w => allSizes.find(p => p >= w) || allSizes[allSizes.length - 1])
```
Comment (hsuanxyz) [inline components/image/image.component.ts:139]:
```suggestion
        [width, width * 2].map(w => allSizes.find(p => p >= w) || w)
```

File: components/image/image.component.ts [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    }
+  }
+
+  ngOnDestroy(): void {
+    this.destroy$.next();
+    this.destroy$.complete();
+  }
+
+  private preload(): void {
+    if (isPlatformServer(this.platformId)) {
```
Comment (hsuanxyz) [inline components/image/image.component.ts:102]: https://material.angular.io/cdk/platform/overview

File: components/image/image.component.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+      linkNode.imageSizes = this.sizes;
+      linkNode.imageSrcset = this.srcSet;
+      linkNode.href = this.src;
+      const headNode = this.document.getElementsByTagName('head')[0] as HTMLHeadElement;
+      this.render.appendChild(headNode, linkNode);
+    }
+  }
+
+  private genSrc(): void {
+    if (!this.nzOptimize) {
```
Comment (hsuanxyz) [inline components/image/image.component.ts:115]: Here need to determine the svg or data

**Comment by stygian-desolator** [comment]
I submitted a new [Pull Request](https://github.com/NG-ZORRO/ng-zorro-antd/pull/6590), this will be closed.

**Pull request closed without merging**